lib/ABI/FunctionType/Layout.cpp

```diff
@@ -568,8 +570,12 @@ ToRawConverter::distributeReturnValue(const QualifiedType &ReturnValue) const {
 }
 
 model::TypePath convertToRaw(const model::CABIFunctionType &Function,
-                             TupleTree<model::Binary> &Binary) {
-  ToRawConverter ToRaw(abi::Definition::get(Function.ABI()));
+                             TupleTree<model::Binary> &Binary,
+                             std::optional<model::ABI::Values> OverrideABI) {
+  if (!OverrideABI.has_value() || *OverrideABI == model::ABI::Invalid)
+    OverrideABI = Function.ABI();
+
+  ToRawConverter ToRaw(abi::Definition::get(OverrideABI.value()));
```

Comment:
This commit adds an ability to explicitly set the ABI even in `CABI->Raw` conversion as a way to ignore whatever ~~broken~~ incorrect stuff is written inside the function itself.

https://github.com/revng/revng/pull/232/commits/a617245555b186e70bbe66d6a853e54bbed116bd

@aleclearmind, please review.

Reply:
The commit is `ABI: let ToRaw conversion accept an ABI override`.

Reply:
This doesn't make sense to me. Overriding the ABI in `CABIFunctionType` is violating the semantics of the `CABIFunctionType`. Why would we do it?

Reply:
I read the comment about `DWARF`, can you please elaborate?

In any case, the situation should either be fixed in `DWARF` or we should change the ABI upstream. This is not the place for this.

Reply:
Dwarf sometimes misdetects ABI pretty severely. For the tests to still be usable, it's either this non-invasive extra option, or a model-pass/tool replacing all the ABIs.

I ~~hate~~ am fine with either option - your call.

Reply:
> Dwarf sometimes misdetects ABI pretty severely

Example?
If it's predictably bad we can forcefully apply the fix directly in `ImportDwarf`.

Reply:
> Example?
`regparm` things.

Reply:
I guess they are not ABIs per se. But we get severely broken stuff when we try to work with those functions as if they were just `SystemV_x86` functions.

Reply:
Sadly, I don't think we can deal with this when importing dwarf. We know for sure in the test case because that's exactly how the function were created. I don't think dwarf carries the kind of information to figure this out looking at an arbitrary function.

Reply:
I don't understand the problem.

Please make a C example, report the corresponding results from `llvm-dwarfdump` and the derived model highlighting what the ABI should be instead of the on it is.

Reply:
Ah, ok, it is indeed an issue with our `DwarfImporter`.

Sadly, it looks like we cannot just ignore `DW_AT_location` (like we do now), because it can lead to incompatible `CABIFunctionType`s :sadparrot:.

If it ever points to a register, we have to label the function with a different ABI in some cases.

Reply:
> Please make a C example, report the corresponding results from `llvm-dwarfdump` and the derived model highlighting what the ABI should be instead of the on it is.

If you'd like feedback from me, please do the above otherwise I'm really not following you.

Reply:
If you find an appropriate place, please leave a TODO about `static` functions whose prototype is available in DWARF but it doesn't follow the standard calling convention.

Reply:
How about [here](https://github.com/revng/revng/blob/develop/lib/Model/Importer/DebugInfo/DwarfImporter.cpp#L202)?

Reply:
Works.

Reply:
Done.